### DISCUSSION TITLE
Extracting interaction effects

### DISCUSSION THREAD

**Discussion by DEHewitt** [bmcclintock/momentuHMM#124] [Q&A] [answered]
Hi @bmcclintock,

I have a quick question about how to extract interaction effects. Specifically, I am interested in the stationary state probabilities for an interaction between two covariates.

If I have the following fitted model:
```
m <- fitHMM(data = prepData,
              nbStates = nbStates, 
              retryFits = retryFits, 
              dist = dist, 
              Par0 = Par0,
              formula = x * y)
```
Where `x` and `y` are two covariates of interest. Would the following be the correct method to obtain the interaction effects? 
```
interaction <- stationary(model = m, covs = data.frame(x = rep(seq(min(x), max(x), 0.01), each = 10), y = rep(seq(min(y), max(y), 0.1), 10))
```
(for arbitrary intervals (i.e., `0.1` for `x` and `0.01` for `y`) of interest throughout the range of each covariate)

In case it helps to clarify, the below is the kind of plot I intend to produce
![image](https://user-images.githubusercontent.com/54158290/153781908-746971af-28eb-4c32-ab56-2cb424fcd0c1.png)

Thanks!

**Comment by bmcclintock** [comment, marked as answer]
I'd suggest using `expand.grid`:

```
library(momentuHMM)
m <- example$m
cov1seq <- seq(min(m$data$cov1),max(m$data$cov1),0.01)
cov2seq <- seq(min(m$data$cov2),max(m$data$cov2),0.1)
interaction <- stationary(model = m, covs = expand.grid(cov1=cov1seq,cov2=cov2seq))
```

**Reply by DEHewitt** [reply]
Thanks for this answer, @bmcclintock (late thanks, but I must've thought marking as answered would suffice). Anyhow, I have come back to this question to see if there is a way to also calculate the SE / CI associated with an interaction term?

Using the pseudo-code from above, I am interested in something like this:
```
cov1levels <- c(2, 6) # assume these are values of interest
cov2seq <- seq(min(m$data$cov2), max(m$data$cov2))

interaction <- stationary(model = m, covs = expand.grid(cov1 = cov1levels, cov2 = cov2seq))
```

Is it possible to calculate the SE / CI along `cov2` for the two supplied levels of `cov1`?

**Reply by bmcclintock** [reply]
It's clunky, but I think you'd need to use `plotStationary` (with `plotCI=TRUE` and `return=TRUE`). Loop over the rows of `covs`  and then create your own table/plot based on the returned SEs and CIs for each row of `covs`.

**Reply by DEHewitt** [reply]
Thanks - that worked! Generic code below so others may adapt if req'd.

```
# Vector of levels of interest
levels <- c(0, 2, 4, 6)

# Define an empty object to store results of the loop
out <- NULL

for (i in 1:length(levels)) {
  
  # Calculate the stationary probabilities 
  stationary <- momentuHMM::plotStationary(model = m, 
                                           covs = data.frame(cov1 = levels[i]), # all other covariates will be at their mean (as per ?plotStationary)
                                           plotCI = TRUE, 
                                           return = TRUE)
  
  # Extract the estimates
  # This could automated using length(m$stateNames) which will save time (esp. where nbStates > 2)
  st1 <- stationary$cov2$`state 1` %>% add_column(state = "State 1")
  st2 <- stationary$cov2$`state 2` %>% add_column(state = "State 2")
  
  # Put them together
  st <- bind_rows(st1, st2)
  
  # Add the level used for these predictions
  st <- st %>% add_column(cov1 = levels[i])
  
  # Add it to the output object
  out <- out %>% bind_rows(st)
  
}
```
